src/mono/mono/metadata/threads.c

```diff
@@ -2060,13 +2060,28 @@ ves_icall_System_Threading_Thread_Join_internal (MonoThreadObjectHandle thread_h
 	return FALSE;
 }
 
+static size_t
+set_pending_null_reference_exception ()
+{
+	ERROR_DECL (error);
+	mono_error_set_null_reference (error);
+	mono_error_set_pending_exception (error);
```

Comment:
I'd appreciate a comment in here to the effect of "this is a bad idea but we're doing it anyway"

In general it's a bad idea to call `mono_error_set_pending_exception` deep in the call graph.  It affects the TLS of a thread and is non-composable and hard to reason about. The recommended pattern is to flow the `error` out to an icall caller and call `mono_error_set_pending_exception` in the icall directly or make the icall `HANDLES` in `icall-def.h` and just let the managed-to-native wrapper deal with it. (not applicable to the interlocked operations, obviously, since they're icalls)